### PR TITLE
protobuf: allow shared builds for Linux

### LINKED ISSUE #3582: [package] protobuf cannot be compiled as shared
Apparently an `if` protecting code supposed to be only for Windows was removed during last PR development and we have this on master:
```
$ conan create . protobuf/3.12.4@ -o protobuf:shared=True
Exporting package recipe
protobuf/3.12.4 exports: File 'conandata.yml' found. Exporting it...
protobuf/3.12.4 exports: Copied 1 '.yml' file: conandata.yml
protobuf/3.12.4 exports_sources: Copied 1 '.txt' file: CMakeLists.txt
protobuf/3.12.4 exports_sources: Copied 2 '.patch' files: upstream-issue-7567-no-export-template-define.patch, upstream-pr-7761-cmake-regex-fix.patch
protobuf/3.12.4: A new conanfile.py version was exported
protobuf/3.12.4: Folder: /home/pfarre/.conan/data/protobuf/3.12.4/_/_/export
protobuf/3.12.4: Exported revision: fcab376e471eb6443f1f3fd012065226
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++
compiler.version=9
os=Linux
os_build=Linux
[options]
protobuf:shared=True
[build_requires]
[env]

ERROR: protobuf/3.12.4: Invalid configuration: On Windows Protobuf can only be built with Visual Studio 2015 or higher.

```

_Originally posted by @Hopobcn in https://github.com/conan-io/conan-center-index/pull/2037#discussion_r524673470_

### PR DESCRIPTION
**Pull request by Croydon**
- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

____

fixes #3582

// @Hopobcn

### CODE REVIEW

**Review by SSE4 — APPROVED** [review]

File: recipes/protobuf/all/conanfile.py [hunk 1]
```diff
@@ -48,15 +48,16 @@ def configure(self):
...
-            compiler_version = Version(self.settings.compiler.version.value)
-            if compiler_version < "14":
-                raise ConanInvalidConfiguration("On Windows Protobuf can only be built with "
-                                                "Visual Studio 2015 or higher.")
-            if self.options.shared and "MT" in self.settings.compiler.runtime:
-                raise ConanInvalidConfiguration("Protobuf can't be build with shared + MT(d) runtimes")
-
-        if tools.is_apple_os(self.settings.os) and self.options.shared:
-            raise ConanInvalidConfiguration("Protobuf could not be build as shared library for Mac.")
+            if self.settings.os == "Windows":
```
Comment (SpaceIm) [inline recipes/protobuf/all/conanfile.py:51]:
```suggestion
            if self.settings.compiler == "Visual Studio":
```

File: recipes/protobuf/all/conanfile.py [hunk 2]
```diff
@@ -48,15 +48,16 @@ def configure(self):
...
+                compiler_version = Version(self.settings.compiler.version.value)
+                if compiler_version < "14":
+                    raise ConanInvalidConfiguration("On Windows Protobuf can only be built with "
+                                                    "Visual Studio 2015 or higher.")
+                if self.options.shared and "MT" in self.settings.compiler.runtime:
+                    raise ConanInvalidConfiguration("Protobuf can't be build with shared + MT(d) runtimes")
+
+            if tools.is_apple_os(self.settings.os):
+                raise ConanInvalidConfiguration("Protobuf could not be build as shared library for Mac.")
```
Comment (SpaceIm) [inline recipes/protobuf/all/conanfile.py:60]:
```suggestion
        if self.settings.compiler == "Visual Studio":
            if Version(self.settings.compiler.version) < "14":
                raise ConanInvalidConfiguration("On Windows Protobuf can only be built with "
                                                "Visual Studio 2015 or higher.")
            if self.options.shared and "MT" in self.settings.compiler.runtime:
                raise ConanInvalidConfiguration("Protobuf can't be build with shared + MT(d) runtimes")

        if tools.is_apple_os(self.settings.os) and self.options.shared:
            raise ConanInvalidConfiguration("Protobuf could not be build as shared library for Mac.")
```

Comment (SpaceIm) [reply recipes/protobuf/all/conanfile.py:60]: this all block shouldn't be under the condition `self.options.shared` deleting fPIC option.

Comment (Croydon) [reply recipes/protobuf/all/conanfile.py:60]: The VS >= 2015 should indeed not be in it, true

Not sure about the runtime. After all Clang for Windows supports those runtimes too, right?

Comment (SpaceIm) [reply recipes/protobuf/all/conanfile.py]: But not MinGW

File: recipes/protobuf/all/conanfile.py [hunk 3]
```diff
@@ -48,16 +48,16 @@ def configure(self):
         if self.options.shared:
             del self.options.fPIC
 
-            compiler_version = Version(self.settings.compiler.version.value)
-            if compiler_version < "14":
-                raise ConanInvalidConfiguration("On Windows Protobuf can only be built with "
-                                                "Visual Studio 2015 or higher.")
-            if self.options.shared and "MT" in self.settings.compiler.runtime:
+            if self.settings.os == "Windows" and "MT" in self.settings.compiler.runtime:
```
Comment (SpaceIm) [inline recipes/protobuf/all/conanfile.py:51]:
```suggestion
            if self.settings.os == "Windows" and self.settings.compiler in ["Visual Studio", "clang"] and "MT" in self.settings.compiler.runtime:
```

File: recipes/protobuf/all/conanfile.py [hunk 4]
```diff
@@ -48,16 +48,16 @@ def configure(self):
...
                 raise ConanInvalidConfiguration("Protobuf can't be build with shared + MT(d) runtimes")
```
Comment (SpaceIm) [inline recipes/protobuf/all/conanfile.py:52]:
```suggestion
                raise ConanInvalidConfiguration("Protobuf can't be built with shared + MT(d) runtimes")
```

File: recipes/protobuf/all/conanfile.py [hunk 5]
```diff
@@ -48,16 +48,16 @@ def configure(self):
...
 
-        if tools.is_apple_os(self.settings.os) and self.options.shared:
-            raise ConanInvalidConfiguration("Protobuf could not be build as shared library for Mac.")
+            if tools.is_apple_os(self.settings.os):
+                raise ConanInvalidConfiguration("Protobuf could not be build as shared library for Mac.")
```
Comment (SpaceIm) [inline recipes/protobuf/all/conanfile.py:55]:
```suggestion
                raise ConanInvalidConfiguration("Protobuf could not be built as shared library for Mac.")
```

**Review by SpaceIm — APPROVED** [review]

**Review by madebr — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**